drift-transport-netty/src/main/java/com/facebook/drift/transport/netty/codec/HeaderTransport.java

```diff
@@ -200,6 +235,94 @@ public static ThriftFrame decodeFrame(ByteBuf buffer)
         }
     }
 
+    private static ByteBuf transform(ByteBuf data, List<Transform> writeTransforms)
+    {
+        if (writeTransforms.contains(ZLIB_TRANSFORM)) {
+            int inputLength = data.readableBytes();
+            byte[] output = new byte[inputLength + ZLIB_COMPRESSION_BUFFER_SIZE]; // output might be larger
+            Deflater compressor = new Deflater();
+            try {
+                byte[] input;
+                int inputOffset;
+                if (data.hasArray()) {
+                    input = data.array();
+                    inputOffset = data.readerIndex();
+                }
+                else {
+                    input = new byte[inputLength];
+                    data.readBytes(input);
+                    inputOffset = 0;
+                }
+                compressor.setInput(input, inputOffset, inputLength);
+                compressor.finish();
+                int length = compressor.deflate(output);
+                if (!compressor.finished()) {
+                    // Output buffer was not big enough.  Unlikely.
+                    // If you hit this, you probably shouldn't be using ZLIB_TRANSFORM :)
+                    throw new IllegalArgumentException("Output compress buffer not big enough");
+                }
+                return Unpooled.wrappedBuffer(output, 0, length);
+            }
+            finally {
+                data.release();
+                compressor.end();
+            }
+        }
+        return data;
+    }
+
+    private static ByteBuf untransform(ByteBuf data, List<Transform> readTransforms)
+    {
+        if (readTransforms.contains(ZLIB_TRANSFORM)) {
+            try {
+                Inflater decompressor = new Inflater();
+                byte[] input;
+                int inputLength = data.readableBytes();
+                int inputOffset;
+                if (data.hasArray()) {
+                    input = data.array();
+                    inputOffset = data.readerIndex();
+                }
+                else {
+                    input = new byte[inputLength];
+                    data.readBytes(input);
+                    inputOffset = 0;
+                }
+                decompressor.setInput(input, inputOffset, inputLength);
+                int length = 0;
+                ArrayList<byte[]> outBytes = new ArrayList<>();
+                while (!decompressor.finished()) {
+                    byte[] output = new byte[ZLIB_COMPRESSION_BUFFER_SIZE];
+                    length += decompressor.inflate(output);
+                    outBytes.add(output);
+                }
+                decompressor.end();
```

Comment:
Any code thrown before this will cause native memory leak. Make sure this is in `finally`.